man/man1/potccd.1
.TH POTCCD 1NEMO "9 January 2022"

.SH NAME
potccd \- Create an image with potentials or densities from a potential

.SH SYNOPSIS
\fBpotccd\fP [parameter=value]

.SH DESCRIPTION
\fIpotccd\fP stores the potential values of a potential descriptor 
(see \fPpotential(5NEMO)\fP) on a regular grid in
standard \fIimage(5NEMO)\fP format. It is also possible to store the
Poissonian density, either 2D or 3D. For this numerical derivatives
of the force field are used, and hence a small difference step needs to
be given.
.PP
A potential stored on a grid could be a good approach when the potential
is expensive to compute, though in orbit calculations this could limit
the exact conservation of the integrals of motion such as energy.
.PP
Irregular grids are not forbidden (see \fBx=,y=,z=\fP) but the
\fIimage(5NEMO)\fP WCS descriptor will then be invalid.

.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBout=\fP
Output file (image). No default.
.TP
\fBpotname=\fP
Name of the \fIpotential(5NEMO)\fP. No default
.TP
\fBpotpars=\fP
Parameters for the potential.
.TP
\fBpotfile=\fP
Any optional data file associated with the potential.
.TP
\fBx=\fP
X-coordinate(s) to test potential at. This should be a regular
array, e.g. 1:10:0.2. Default: 0.
.TP
\fBy=\fP
Y-coordinate(s) to test potential at. Default: 0.
.TP
\fBz=\fP
Z-coordinate(s) to test potential at. Default: 0.
.TP
\fBt=\fP
Time to test potential at   
.TP
\fBmode=pot|ax|ay|az|ar|at|den\fP
Output mode. Choices are potential, accelerations in X, Y or Z, and
density (which needs dr>0).
Default: \fBpot\fP.
.TP
\fBdr=\fP
Difference step used to compute numerical force derivates that
are used to compute Poissonian densities. 
.TP
\fBomega=\fP
Use this instead of any returned pattern speed. Default: not used, the first
parameter in \fBpotpars=\fP is used.
.TP
\fBndim=\fP
Number of dimensions used in Poissonian density computation. Should
be 2 or 3. 
.TP
\fBnder=1|2\fP
Normally the density (\fBmode=den\fP) is derived by first derivatives
of the forces, but for some potentials the force was never implemented
(e.g. \fBpotname=gauss\fP).  In this case the 2nd derivative of the
potential will be needed, hence \fBnder=2\fP.
Default: 1

.SH BUGS
With \fBndim=2\fP the Z-coordinate is not ignored, and hence may
give meaningless results if forces depend on Z.

.SH EXAMPLES
First a somewhat obscure way in to create a map of an arbitrary function
using the \fBpotname=rotcur\fP lookup table. It will use
a spline interpolation, and if \fBmode=pot\fP is used, the "potential"
is actually the "rotation curve" itself, since in general one cannot
easily compute the potential from a rotation curve without knowing
the full geometry. Here is an example how to create a smooth
radial profile of the function "f(x)=1/sqrt(4+x)" on a grid from
-10..10:
.nf
    % nemoinp 0:10 | tabmath - - '1/(sqrt(4+%1)' > map0.tab
    % potccd map0 rotcur 0 map0.tab x=-10:10:0.1 y=-10:10:0.1
.fi
If the functional form is know, \fIccdmath(1NEMO)\fP will perhaps
do better, if not a little more involved to type.
.PP
Here is an example of an orbit integration in the analytical plummer potential,
and in two gridded versions of the potential.

.nf
% mkorbit o1 1 1 0 -0.5 0 0 potname=plummer
etot: -0.452350

% orbint o1 - 10000 0.01 ndiag=10000 | orblist - | tail -1
Energy conservation: 2.05084e-11
10000 100 -1.11384 -0.266207 0 0.557843 -0.315574 0

% potccd ccd1 plummer x=-2:2:0.01 y=-2:2:0.01
% orbint o1 - 10000 0.01 ndiag=10000 potname=ccd potfile=ccd1 | orblist - | tail -1
Energy conservation: 0.00223468
10000 100 -1.11427 -0.265478 0 0.557503 -0.315907 0 

% potccd ccd2 plummer x=-2:2:0.0025 y=-2:2:0.0025
% orbint o1 - 10000 0.01 ndiag=10000 potname=ccd potfile=ccd2 | orblist - | tail -1
Energy conservation: 0.000609561
10000 100 -1.11385 -0.266169 0 0.557829 -0.315593 0

.fi

.SH SEE ALSO
potlist(1NEMO), image(5NEMO), potential(3NEMO), potential(5NEMO)

.SH AUTHOR
Peter Teuben

.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
10-Jun-92	V1.0 Created       	PJT
30-mar-94	V1.1 added density options (dr=, ndim=)	PJT
12-sep-02	V1.2 added mode=	PJT
19-mar-2021	V2.1 added nder=	PJT
9-jan-2022	V2.2 fixed WCS, updated examples	PJT
.fi


src/orbit/potential/potccd.c
/*
 * POTCCD: Turn potential into an (potential|density) image;
 *	   useful for potname=ccd
 *
 *	10-jun-92	Created		        Peter Teuben
 *	30-mar-94	poisson density		pjt
 *	 2-jul-95	documented bug fix on rotating potentials 	pjt
 *      19-apr-96       improved error reporting        pjt
 *      12-jun-98       made ndim=3 the default, like potlist           pjt
 *	13-sep-01	better prototype for proc			pjt
 *       4-dec-01       also compute min/max
 *      12-sep-02       optionally output a force                       pjt
 *	22-oct-02       also allow ar,at                                pjt
 *      16-mar-2021     axis=1 proper WCS                               PJT
 *       9-jan-2022     fix Xref for axis=1                             pjt
 */

#include <stdinc.h>
#include <getparam.h>
#include <image.h>
#include <potential.h>

string defv[] = {
    "out=???\n      Output file (image)",
    "potname=???\n  Name of potential",
    "potpars=\n     Parameters for potential",
    "potfile=\n     Any optional data file associated with potential",
    "x=0\n          X-coordinate(s) to test potential at",
    "y=0\n          Y-coordinate(s) to test potential at",
    "z=0\n          Z-coordinate(s) to test potential at",
    "t=0.0\n        Time to test potential at",
    "mode=pot\n     Output pot,ax,ay,az,ar,at,den",
    "dr=\n          Differential step for (Poisson) density map",
    "omega=\n       Use this instead of any returned pattern speed",
    "ndim=3\n       Poisson map using 2D or 3D derivatives",
    "nder=1\n       1: use force der   2: use twice potential der for density",
    "VERSION=2.2a\n 9-jan-2022 PJT",
    NULL,
};

string usage = "Create potential or density image from a NEMO potential";

/*                     this code uses double[3*MAXPT] */
#ifndef MAXPT
# define MAXPT 10000
#endif

local potproc_double mypot;    /* pointer to potential calculator function */

void nemo_main(void)
{
    int    nx,ny,nz, ix,iy,iz;
    double pos[3],acc[3],pot,den,dr,time;
    double xarr[MAXPT],yarr[MAXPT],zarr[MAXPT];
    double omega, dmin=0, dmax=0;
    string mode = getparam("mode");
    int idim, ndim, maxdim = 3, first=1, idx = 0;
    int axis = 0; // new WCS system
    int nder = getiparam("nder");
    imageptr iptr;
    stream ostr;

    ostr = stropen(getparam("out"),"w");
    nx = nemoinpd(getparam("x"), xarr, MAXPT);  /* get sample arrays */
    ny = nemoinpd(getparam("y"), yarr, MAXPT);
    nz = nemoinpd(getparam("z"), zarr, MAXPT);
    dprintf(0,"Creating image %d * %d * %d\n",nx,ny,nz);
    if (nx > 1 || ny > 1 || nz > 1) {  /* check if > 1 */
        if (nx > 1) {
           if (ny>1 && ny!=nx) error("ny <> nx\n");
           if (nz>1 && nz!=nx) error("nz <> nx\n");
        } else if (ny > 1) {
           if (nz>1 && nz!=ny) error("nz <> ny\n");
        } 
    } 
    if (nx < 0 || ny < 0 || nz < 0)
        error("problem with your grid, check your x=,y=,z=");
    ndim = getiparam("ndim");
    if (hasvalue("dr")) {
        dr = getdparam("dr");
        dprintf(0,"%dD-Poisson density map with delta-r=%g and nder=%d\n",ndim,dr,nder);
    } else
        dr = -1.0;
    time = getdparam("t");
    if (streq(mode,"pot"))
      idx = 0;
    else if (streq(mode,"ax"))
      idx = 1;
    else if (streq(mode,"ay"))
      idx = 2;
    else if (streq(mode,"az"))
      idx = 3;
    else if (streq(mode,"ar"))
      idx = 4;
    else if (streq(mode,"at"))
      idx = 5;
    else if (streq(mode,"den")) {
      idx = 0;
      if (dr < 0) error("Need to supply a small positive value for dr=");
    } else
      error("bad mode=%s; allowed are: ax,ay,az,ar,at,den",mode);
    mypot = get_potential(getparam("potname"), 
			  getparam("potpars"), 
			  getparam("potfile"));
			  
    if (mypot==NULL) error("Potential could not be loaded");
    if (hasvalue("omega"))
      omega = getdparam("omega");
    else
      omega = get_pattern();
    dprintf(0,"using Omega = %g\n",omega);

    create_cube(&iptr,nx,ny,nz);
    Dx(iptr) = (nx > 1 ? xarr[1]-xarr[0] : 0.0);
    Dy(iptr) = (ny > 1 ? yarr[1]-yarr[0] : 0.0);
    Dz(iptr) = (nz > 1 ? zarr[1]-zarr[0] : 0.0);

    axis = 1;
    Axis(iptr) = axis;
    if (axis == 0) {        // deprecated
      Xmin(iptr) = xarr[0];
      Ymin(iptr) = yarr[0];
      Zmin(iptr) = zarr[0];
      Xref(iptr) = 0.0;
      Yref(iptr) = 0.0;
      Zref(iptr) = 0.0;
    } else {
      dprintf(1,"X:  %g .. %g  -> %g\n", xarr[0], xarr[nx-1], -xarr[0]/Dx(iptr));
      dprintf(1,"Y:  %g .. %g  -> %g\n", yarr[0], yarr[nx-1], -yarr[0]/Dy(iptr));
      Xmin(iptr) = 0.0;  // force the reference pixel at (0,0)
      Ymin(iptr) = 0.0;
      Zmin(iptr) = 0.0;
      Xref(iptr) = -xarr[0]/Dx(iptr);
      Yref(iptr) = -yarr[0]/Dy(iptr);
      Zref(iptr) = 0.0;
    }

    for (iz=0; iz<nz; iz++) {
      pos[2] = zarr[iz];
      for (iy=0; iy<ny; iy++) {
	pos[1] = yarr[iy];
	for (ix=0; ix<nx; ix++) {
	  pos[0] = xarr[ix];
	  (*mypot)(&maxdim,pos,acc,&pot,&time);
	  if (dr > 0.0) {                                 /* Poisson */
	    if (nder==1) {
	      den = 0.0;
	      for (idim=0; idim<ndim; idim++) {
		pos[idim] += dr;
		(*mypot)(&maxdim,pos,acc,&pot,&time);
		den -= acc[idim];
		pos[idim] -= 2*dr;
		(*mypot)(&maxdim,pos,acc,&pot,&time);
		den += acc[idim];
		pos[idim] += dr;
	      }
	    } else {
	      den = -2*ndim*pot;
	      for (idim=0; idim<ndim; idim++) {
		pos[idim] += dr;
		(*mypot)(&maxdim,pos,acc,&pot,&time);
		den += pot;
		pos[idim] -= 2*dr;
		(*mypot)(&maxdim,pos,acc,&pot,&time);
		den += pot;
		pos[idim] += dr;
	      }
	      den = den/dr;
	    }
	    pot = den/dr;
	  } else if (idx==0) {                                        /* Potential */
	    if (omega != 0.0) {
	      pot -= 0.5*sqr(omega)*(sqr(pos[0])+sqr(pos[1]));
	    }
	  } else if (idx < 4) {
	    pot = acc[idx-1];
	  } else {
	    /* 2D only */
	    real vv,vr,rr;
	    vv = acc[0]*acc[0] + acc[1]*acc[1];
	    vr = acc[0]*pos[0] + acc[1]*pos[1];
	    rr = pos[0]*pos[0] + pos[1]*pos[1];
	    if (idx == 4)
	      pot = vr/sqrt(rr);
	    else
	      pot = sqrt(vv-vr*vr/rr);
	  }
	  CubeValue(iptr,ix,iy,iz) = pot;
	  if (first) {
	    dmin = dmax = pot;
	    first = 0;
	  } else {
	    dmin = MIN(dmin, pot);
	    dmax = MAX(dmax, pot);
	  }
	}
      }
    }
    MapMin(iptr) = dmin;
    MapMax(iptr) = dmax;
    write_image(ostr, iptr);
}
